torch/_dynamo/variables/torch.py

```diff
@@ -738,6 +738,86 @@ def handle_ntuple(value):
             return handle_ntuple(args[0])
 
 
+def is_fn_safe_to_run(tx, f, sub_args):
+    # Snapshot state
+    graph_checkpoint, checkpoint = tx.output.graph, tx.copy_graphstate()
+    # Will raise if not sound
+    output, graph, lifted_freevars = speculate_subgraph(
+        tx, f, sub_args, graph_checkpoint, checkpoint, always_restore=True
+    )
+    # If we got here, we are probably sound, but we do not support freevars yet, so lets call cases where we found them
+    # unsafe.
+    if lifted_freevars:
+        return False
+    return True
+
+
+# See NOTE [HigherOrderOperator tracing design] for details of the design
+# See NOTE [speculate_subgraph vs old_speculate_subgraph] for other info
+def speculate_subgraph(
+    tx, f, sub_args, graph_checkpoint, checkpoint, *, always_restore=False
```

Comment:
Is always_restore the worst named variable ever? Maybe we should flip this to speculate_only? or refactor into 2 fns of speculation (arg building + invocation + asserts) from installation onto tx?

Reply:
`always_restore` or `speculate_only` sound fine to me.

speculate_subgraph was originally rewritten in the second style (see old_speculate_subgraph): it would do "arg building + invocation + asserts", and then the control flow op (cond, map) ends up installing the guards and/or side effects. But it was difficult to get a clean split, because as a part of speculating the subgraph, we actually modify the original graph with things like new graph inputs, so it is now in the form that you see.